### ISSUE TITLE
typemap(in) fails to match structure member

### ISSUE THREAD

**Issue by jschueller** [swig/swig#2056]
Lets define a simple struct:
```
typedef struct _A {
  int *v;
} A;
```

Now I can define a typemap(out) for the specific member v inside struct A:
```
%typemap(out) int * _A::v {
  ...
```

But if I want to define a typemap(in) in the same way it is not matched:
```
%typemap(in) int * A::v {
...
```

A similar typemap without the struct specifier works:
```
%typemap(in) int * v {
```
but I'm stuck because I'd like to write different typemaps for members from different structs but with the same name.

The question is do typemap(out) and typemap(in) would differ here ?

I put the files to reproduce here: https://github.com/jschueller/swig-typemap-struct

**Comment by jschueller** [comment]
using the swig option -debug-tmused I can see that no typemap in is issued for _A::v, only the out typemap is matched:
```
example.h:4: Typemap for int *_A::v (out) : %typemap(out) int *_A::v
```

**Comment by jschueller** [comment]
maybe @vadz knows about typemaps emissions ?

**Comment by jschueller** [comment]
maybe cc @LindleyF (I saw you just submited a PR for typemaps) ?

**Comment by vadz** [comment]
Maybe try using `-debug-tmsearch` instead? This will give (much) more output, but there could be something useful in it.

Also, is this Python-specific?

**Comment by jschueller** [comment]
Indeed -debug-tmsearch is very useful: it does not look for an 'in' typemap for _A::v, only for the out typemap
```
$ make | grep "Searching for a suitable" |grep v
example.h:4: Searching for a suitable 'memberin' typemap for: int *v
example.h:4: Searching for a suitable 'default' typemap for: int *v
example.h:4: Searching for a suitable 'arginit' typemap for: int *v
example.h:4: Searching for a suitable 'in' typemap for: int *v
example.h:4: Searching for a suitable 'typecheck' typemap for: int *v
example.h:4: Searching for a suitable 'argout' typemap for: int *v
example.h:4: Searching for a suitable 'check' typemap for: int *v
example.h:4: Searching for a suitable 'freearg' typemap for: int *v
example.h:4: Searching for a suitable 'except' typemap for: void _A::v
example.h:4: Searching for a suitable 'out' typemap for: void _A::v
example.h:4: Searching for a suitable 'ret' typemap for: void _A::v
example.h:4: Searching for a suitable 'except' typemap for: int *_A::v
example.h:4: Searching for a suitable 'out' typemap for: int *_A::v
example.h:4: Searching for a suitable 'ret' typemap for: int *_A::v
example.h:3: Searching for a suitable 'except' typemap for: void _A::~_A
example.h:3: Searching for a suitable 'out' typemap for: void _A::~_A
example.h:3: Searching for a suitable 'ret' typemap for: void _A::~_A
example.h:10: Searching for a suitable 'except' typemap for: void init_a
example.h:10: Searching for a suitable 'out' typemap for: void init_a
example.h:10: Searching for a suitable 'ret' typemap for: void init_a
```

**Comment by jschueller** [comment]
Indeed, it does not seem to be Python-specific, I tried ruby and it has the same behavior

**Comment by jschueller** [comment]
In functionWrapper, I can replace Swig_typemap_lookup_out(...)  by Swig_typemap_lookup("in", ...)
to emit some typemap(in) for int * _A::v (as seen with -debug-tmsearch) but the wrapper compilation fails (also I really dont know what I'm doing)
```
diff --git a/Source/Modules/python.cxx b/Source/Modules/python.cxx
index 6bc7ef737..71970127c 100644
--- a/Source/Modules/python.cxx
+++ b/Source/Modules/python.cxx
@@ -3084,7 +3084,8 @@ public:
     /* This part below still needs cleanup */
 
     /* Return the function value */
-    tm = Swig_typemap_lookup_out("out", n, Swig_cresult_name(), f, actioncode);
+//    tm = Swig_typemap_lookup_out("out", n, Swig_cresult_name(), f, actioncode);
+tm = Swig_typemap_lookup("in", n, Swig_cresult_name(), f);
```

**Comment by jschueller** [comment]
turns out functionWrapper does not call Swig_typemap_lookup for input arguments:
https://github.com/swig/swig/blob/master/Source/Modules/python.cxx#L2860

It seems it's the Swig_typemap_lookup function that has the qualified name search logic:
https://github.com/swig/swig/blob/master/Source/Swig/typemap.c#L1325

I dont know whether functionWrapper should call Swig_typemap_lookup or mimic the qualified name search somehow, any ideas @vadz ?

**Comment by vadz** [comment]
Sorry, I have no idea, I'm afraid William is the only to really know this code.

**Comment by ojwb** [comment]
Still reproducible with current git master.  Sadly this isn't an area I know much about though.

**Comment by wsfulton** [comment]
According to the [Structure data members](https://swig.org/Doc4.3/SWIG.html#SWIG_structure_data_members) docs, setting a member is akin to wrapping a regular function, so in this case it is similar to
```
void A_v_set(A *a, int *v) {
  a->v = v;
}
```
SWIG has no way to target the `int *v` in this function compared to another function with a parameter of the same name. This is the case for regular functions where there is no C/C++ way to scope the v variable. Quite possibly SWIG could be enhanced as a special case for member variables to additionally perform a typemap lookup for `A::v` as this scoping is arguably valid.

**Comment by jschueller** [comment]
do you have a clue why the `A::v` scoping is valid for the output typemap but not for the input typemap ? it seems asymetric in that regard

**Comment by wsfulton** [comment]
It is assymetric yes. However, the typemap rules are documented for a member getter method such that `A::v` is a valid scoping for a method `v` in `A`. It's based on the fact that in C++, `int * A::v()` is a valid member (akin to the getter function for member variable wrappers) and the scoping for the type in this case is `int *A::v`.

**Comment by jschueller** [comment]
ok, thanks

**Comment by ojwb** [comment]
> SWIG has no way to target the `int *v` in this function compared to another function with a parameter of the same name.

That's also a problematic limitation sometimes.  I actually ended up renaming some parameters in Xapian's C++ API to work around this, but that's not a good solution in general as it relies on being able to convince the maintainers of the library you're wrapping to accept such changes.

As you say, there's not a C/C++ syntax to refer to a parameter name in a specific function (because that's not needed in C/C++).  I don't have a good proposal for how this could be specified.

I think it would be good to at least explicitly note this asymmetry in how SWIG matches `in`/`out` typemaps against member variables in the docs you link to.  It seems surprising until you understand the logic behind it (actually, it still seems surprising even now I do understand it!)
